src/evaluate/typed.ts

```diff
@@ -0,0 +1,107 @@
+import * as reflect from 'jsii-reflect';
+import { TemplateExpression } from '../parser/template';
+import { EvaluationContext, EvaluationContextOptions } from './context';
+import { Evaluator } from './evaluate';
+
+export interface TypedContextOptions extends EvaluationContextOptions {
+  readonly typeSystem: reflect.TypeSystem;
+}
+
+export class TypedContext<P, A = P> extends EvaluationContext<P, A> {
+  public readonly typeSystem: reflect.TypeSystem;
+
+  constructor(opts: TypedContextOptions) {
+    super(opts);
+    this.typeSystem = opts.typeSystem;
+  }
+
+  /**
+   * Return the Class for a given FQN
+   */
+  public resolveClass(fqn: string) {
+    const [mod, ...className] = fqn.split('.');
+    // eslint-disable-next-line @typescript-eslint/no-require-imports
+    const module = require(mod);
+
+    let curr = module;
+    while (true) {
+      const next = className.shift();
+      if (!next) {
+        break;
+      }
+      curr = curr[next];
+      if (!curr) {
+        throw new Error(`unable to resolve class ${className}`);
+      }
+    }
+
+    return curr;
+  }
+
+  // @todo make it generic
+  public extractPropsType(fqn: string): reflect.TypeReference | undefined {
+    const construct = this.typeSystem.findFqn(fqn);
+    if (!construct.isClassType()) {
+      return;
+    }
+
+    const [_scopeParam, _idParam, propsParam] =
+      construct.initializer?.parameters ?? [];
+
+    return propsParam?.type;
+  }
+}
+
+enum EvaluationType {
+  UNKNOWN,
+  VOID,
+  ANY,
+  NUMBER,
+  BOOLEAN,
+  STRING,
+  ARRAY_OF_TYPE,
+  MAP_OF_TYPE,
+  UNION_OF_TYPE,
+  ENUM,
+  ENUM_LIKE,
+  INTERFACE,
+  STATIC_PROPERTY,
+  STATIC_METHOD,
+}
+
+export abstract class TypedEvaluator<
+  Context extends EvaluationContext<unknown, unknown>,
+  Resource
+> extends Evaluator<Context, Resource> {
+  public evaluateTyped(x: TemplateExpression, typeRef?: reflect.TypeReference) {
+    if (typeRef === undefined) {
+      return undefined;
+    }
+
+    switch (this.detectEvaluationType(typeRef)) {
+      default:
+        return this.evaluate(x);
+    }
+  }
+
+  public evaluateObjectTyped(
+    fields: Record<string, TemplateExpression>,
+    typeRef?: reflect.TypeReference
+  ) {
+    return this.evaluateTyped({ type: 'object', fields }, typeRef);
+  }
+
+  private detectEvaluationType(typeRef: reflect.TypeReference): EvaluationType {
+    if (typeRef.primitive === 'number') {
```

Comment:
And then, if it's a primitive, the evaluation type would be something like `NO_OP`. There is nothing else to do.